replication.go

```diff
@@ -0,0 +1,201 @@
+		e.handleFinalizationCertificateResponse(resp.FinalizationCertificateResponse, from)
+	}
+	if resp.LatestRoundResponse != nil {
+		e.handleLatestRoundResponse(resp.LatestRoundResponse)
+	}
+}
+
+func (e *Epoch) handleFinalizationCertificateResponse(resp *FinalizationCertificateResponse, from NodeID) error {
+	e.Logger.Debug("Received finalization certificate response", zap.String("from", from.String()), zap.Uint64("seq", resp.FCert.Finalization.Seq))
+	if e.round+e.maxRoundWindow < resp.FCert.Finalization.Seq {
+		// we are too far behind, we should ignore this message
+		return nil
+	}
+
+	// if we already have a round object for this round, we should persist the fCert
+	round, ok := e.rounds[resp.Block.BlockHeader().Round]
+	if !ok {
+		e.storeFutureFinalizationResponse(resp, from)
+		return nil
+	}
+	if round.fCert != nil {
+		return nil
+	}
+	err := e.persistFinalizationCertificate(resp.FCert)
+	if err != nil {
+		return err
+	}
+
+	if e.round == resp.FCert.Finalization.Seq {
+		e.increaseRound()
+	}
+
+	// handle future messages in case we need to persist more fCerts from the future
+	return e.maybeLoadFutureMessages(e.round)
+}
+
+func (e *Epoch) handleLatestRoundResponse(r *LatestRoundResponse) {
+	// if we get a round later than ours, save it fCertReplicationState
+	if r.Block.BlockHeader().Round > e.latestRoundKnown.num {
+		e.latestRoundKnown.num = r.Block.BlockHeader().Round
+		e.latestRoundKnown.block = r.Block
+		e.latestRoundKnown.notarization = r.Notarization
+		e.latestRoundKnown.fCert = r.FCert
+	}
+}
+
+func (e *Epoch) storeFutureFinalizationResponse(resp *FinalizationCertificateResponse, from NodeID) {
+	msg, ok := e.futureMessages[string(from)][resp.Block.BlockHeader().Round]
+	if !ok {
+		msgsForRound := &messagesForRound{}
+		msgsForRound.proposal = &BlockMessage{
+			Block: resp.Block,
+		}
+		e.futureMessages[string(from)][resp.Block.BlockHeader().Round] = msgsForRound
+		return
+	}
+
+	if msg.proposal != nil && bytes.Equal(msg.proposal.Block.Bytes(), resp.Block.Bytes()) {
+		e.Logger.Error("Proposal does not match the block in the finalization certificate response", zap.String("from", from.String()))
+		return
+	}
+	msg.fCert = &resp.FCert
+	msg.proposal.Block = resp.Block
+}
+
+// SetLastReceivedFCertSeq updates the last received finalization certificate sequence number
+// if [seq] is greater than the current last received sequence number
+func (e *Epoch) setLastReceivedFCertSeq(seq uint64) {
+	if seq > e.latestRoundKnown.num {
+		e.latestRoundKnown.num = seq
+		e.latestRoundKnown.block = nil
+		e.latestRoundKnown.notarization = nil
+		e.latestRoundKnown.fCert = nil
+	}
+}
+
+func (e *Epoch) sendFutureCertficatesRequests(start uint64, end uint64, comm Communication) {
+	if e.lastSequenceRequested >= end {
+		// no need to resend
+		return
+	}
+
+	// we want to collect
+	for seq := start; seq <= end; seq++ {
+		// also request latest round in case this fCert is also behind
```

Comment:
if we always ask for the latest round, then why not just always include that in the response?